### PR TITLE
Setup lint to the future

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
_Reviewers:_ you'll want to review per-commit.
_Supersedes_: https://github.com/discourse/discourse/pull/20844

## How does this work?

Any time a lint rule is added or changed, you can run `yarn lint:fix` to handle all the auto-fixable situations.
But not all lints are auto-fixable -- for those, lint-to-the-future has tooling to automatically ignore present violations.
An alias has been added for lint-to-the-future to ignore new violations, `yarn lttf:ignore`.
The command will add lint-ignore declarations throughout all the files with present violations, which should then be committed.

An excerpt from lint-to-the-future's [README](https://github.com/mansona/lint-to-the-future#lint-to-the-future-dashboard):

> The point of Lint to the Future is to allow you to progressively update your codebase using new lint rules without overwhelming you with the task. You can easily ignore lint rules using project-based ignores in your config files but that doesn't prevent you from making the same errors in new files.

> We chose to do the ignores on a file basis as it is a perfect balance and it means that the tracking/graphing aspects of Lint to the Future provide you with achievable goals, especially in large codebases.

## How do I view progress?

lint-to-the-future provides graphs of violations-over-time per lint rule in a dashboard format, so we can track how well we're doing at cleaning up the violations.

This PR adds a github workflow which will deploy the dashboard to github-pages on the `gh-pages` branch. 
Once GitHub Pages are enabled, and configured to use the `gh-pages` branch, and after this PR is merged, the dashboard will be available for viewing at https://discourse.github.io/discourse
If this is not desired, there are other [deployment options](https://github.com/mansona/lint-to-the-future/blob/main/docs/dashboard-deployment.md) -- for example, cloudflare-pages could be used with a custom URL.
It's also perfectly fine to _not_ deploy the dashboard somewhere, if that is desired.

To view the dashboard locally, run `yarn lint-progress` and visit `http://localhost:8084` (or whatever the port it chose, as it will choose a new port if 8084 is preoccupied)

Also there is a `list` command which shows a JSON object of:
```ts
{
  [date: string]: { // yyyy-mm-dd
    [pluginName: string]: {
      [fileName: string]: string[]; // list of files with violations
    }
  }
}
```

```bash
yarn lint-to-the-future list --stdout
```

## Notes for review

- view each commit separately
- due to dependency changes to improve dependency resolution stability, the repo is was in an invalid state. 
- I've used the lint-to-the-future system to manage the new lint violations
- lint-to-the-future-ember-templates have a bug, which I've reported here: https://github.com/mansona/lint-to-the-future-ember-template/issues/21 (and PR'd)
- To verify locally, run `yarn lint` -- there should be 0 errors and exit status 0. Pop off a couple commits, and there will be errors.

## What about lint-todo?

Lint todo is another system available for both eslint and ember-template-lint that _forces_ folks to "leave things better than they found them" by being transparent / line-specific ignoring of violations. 
It was decided that for _this_ project, it made more sense, and would be less disruptive to new contributors to have the ignore declarations explicitly defined in each file (whereas in lint-todo, they are hidden).
To effectively use lint-todo, a whole team needs to agree to the workflow, and in open source, we want "just anyway" to be able to contribute, and throwing surprises at them can deter contributions.

### CODE REVIEW

File: app/assets/javascripts/admin/addon/templates/user-fields.hbs [hunk 1]
```diff
@@ -8,7 +8,7 @@
       <AdminUserFieldItem
         @userField={{uf}}
         @fieldTypes={{this.fieldTypes}}
-        @firstField={{this.sortedFields.firstObject}}
```
Comment (davidtaylorhq) [inline app/assets/javascripts/admin/addon/templates/user-fields.hbs:11]: I'm a little confused about why these changes were made - there were no violations before, and I don't think this PR is introducing any linting rules? 🤔

File: app/assets/javascripts/admin/addon/templates/user-fields.hbs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+{{! template-lint-disable no-array-prototype-extensions }}
```
Comment (davidtaylorhq) [inline app/assets/javascripts/admin/addon/templates/user-fields.hbs:1]: Similarly... did we introduce some new rules? This wasn't causing any violation before 🤔

**Comment by NullVoxPopuli** [comment]
>  Also, please can you update the linting GitHub actions workflow to use these new yarn lint:... shortcuts?

added!

**Review by davidtaylorhq — APPROVED** [review]
